Improve documentation on iterators length

LeSeulArtichaut opened the pull request:
Attempts to resolve #66491. @the8472 does this help?

r? @steveklabnik

steveklabnik commented on src/libcore/iter/traits/exact_size.rs:
```diff
@@ -69,7 +69,8 @@
 pub trait ExactSizeIterator: Iterator {
-    /// Returns the exact number of times the iterator will iterate.
+    /// Returns the exact length of the iterator, which is the number of times
+    /// the iterator will return `Some(T)` before returning `None`.
```
This is a *little* long for a summary sentence. I'm not sure of the best way to re-word it though. It's likely that some of this will have to go below.

LeSeulArtichaut replied:
What about:
```markdown
Returns the exact length of the iterator.

The implementation ensures that the iterator will return exactly `len()` times a `Some(T)` value, before returning `None`.

the8472 commented:
Seems ok. Maybe instead "which is the *remaining* number of times" would better convey that is reflects the current state of the iterator.

JohnCSimon commented:
Ping from triage:
@LeSeulArtichaut can you please address the comments? Thank you.

LeSeulArtichaut commented:
Addressed the review comments. @steveklabnik will you have time to review those changes? Thank you!

@rustbot modify labels: -S-waiting-on-author +S-waiting-on-review

steveklabnik commented:
Looks good to me, thank you!

@bors: r+ rollup

bors commented:
:pushpin: Commit 0c82a5c1d35c84859ed2f88194858adae831d06d has been approved by `steveklabnik`

<!-- @bors r=steveklabnik 0c82a5c1d35c84859ed2f88194858adae831d06d -->
<!-- homu: {"type":"Approved","sha":"0c82a5c1d35c84859ed2f88194858adae831d06d","approver":"steveklabnik"} -->
